gitlab/resource_gitlab_group.go

```diff
@@ -117,8 +118,13 @@ func resourceGitlabGroupRead(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*gitlab.Client)
 	log.Printf("[DEBUG] read gitlab group %s", d.Id())
 
-	group, _, err := client.Groups.GetGroup(d.Id())
+	group, resp, err := client.Groups.GetGroup(d.Id())
 	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			log.Printf("[DEBUG] gitlab group %s not found so removing from state", d.Id())
```

Comment:
Sometimes a 404 can be due to the member losing access to the group, and not because a group is no longer there. I think only the user can truly tell if a GitLab Group is no longer there or if they just lost access. In the former's case, I think it is best to let the user manually remove the group from the state themselves. In a team-based environment where this resource is shared, this change would remove the group from the state accidentally while others still had access to it.

Reply:
@praneetloke the normal convention is that if something is manually deleted then the TF provider should remove it from state. 404 means Not around - if it’s because of not access then I’d expect a different StatusCode

This is the correct fix and should actually be applied across all of the resources eventually IMO. This is how the aws provider works as well

Reply:
> if it’s because of not access then I’d expect a different StatusCode

GitLab API returns 404 for groups that a user may not have access to. GitHub API does the same thing, too. It is similar to why when you try to access a private repo that you know exists, you receive a 404 instead of a 403 Forbidden.

Reply:
So the question is, does the message that is returned any different between an access issue and a deleted repo? If not, that seems like a major upstream error in Gitlab.

Reply:
> So the question is, does the message that is returned any different between an access issue and a deleted repo?

Nope. That's what I am trying to say :) But you can confirm this easily with the API.

Reply:
@stack72 the ambiguity of 404 in gitlab api's made us remove the 404 handling...

But I think lots of people complain the other way around.

I would be OK to re-add it for resources that support import.

Reply:
Maybe a config option to let users pick if a 404 should be treated as a deleted resource? It would be global to this provider, of course and each resource that his behavior would have to honor that, of course.